### ISSUE TITLE
Tracker: cupyx.scipy.signal

### ISSUE THREAD

**Issue by andfoy** [cupy/cupy#7403]
## Description

This is a tracking issue regarding the current and future state of compatibility between `scipy.signal` and `cupyx.scipy.signal`. The sections denoted below are aligned with respect to SciPy's reference API: https://docs.scipy.org/doc/scipy/reference/signal.html

---

### Convolution
- [x] `convolve` → https://github.com/cupy/cupy/pull/3748
- [x] `correlate` → https://github.com/cupy/cupy/pull/3748
- [x] `fftconvolve` → https://github.com/cupy/cupy/pull/3828
- [x] `oaconvolve` → https://github.com/cupy/cupy/pull/4468
- [x] `convolve2d` → https://github.com/cupy/cupy/pull/3748
- [x] `correlate2d` → https://github.com/cupy/cupy/pull/3748
- [x] `sepfir2d` → https://github.com/cupy/cupy/pull/3645
- [x] `choose_conv_method` → https://github.com/cupy/cupy/pull/3464
- [ ] `correlation_lags`

---

### B-Splines
- [ ] `bspline`
- [ ] `cubic`
- [ ] `quadratic`
- [ ] `gauss_spline`
- [ ] `cspline1d`
- [ ] `qspline1d`
- [ ] `cspline2d`
- [ ] `qspline2d`
- [ ] `cspline1d_eval`
- [ ] `qspline1d_eval`
- [ ] `spline_filter`

---

### Filtering
- [x] `order_filter`  → https://github.com/cupy/cupy/pull/3645
- [x] `medfilt` → https://github.com/cupy/cupy/pull/3645
- [x] `medfilt2d` → https://github.com/cupy/cupy/pull/3645
- [x] `wiener` → https://github.com/cupy/cupy/pull/3645
- [ ] `symiirorder1`
- [ ] `symiirorder2`
- [ ] `lfilter`
- [ ] `lfiltic`
- [ ] `lfilter_zi`
- [ ] `filtfilt`
- [ ] `savgol_filter`
- [ ] `deconvolve`
- [ ] `sosfilt`
- [ ] `sosfilt_zi`
- [ ] `sosfiltfilt`
- [ ] `hilbert`
- [ ] `hilbert2`
- [ ] `decimate`
- [ ] `detrend`
- [ ] `resample`
- [ ] `resample_poly`
- [ ] `upfirdn` 

---

### Filter design
- [ ] `bilinear`
- [ ] `bilinear_zpk`
- [ ] `findfreqs`
- [ ] `firls`
- [ ] `firwin`
- [ ] `firwin2`
- [ ] `freqs`
- [ ] `freqs_zpk`
- [ ] `freqz`
- [ ] `freqz_zpk`
- [ ] `sosfreqz`
- [ ] `gammatone`
- [ ] `group_delay`
- [ ] `iirdesign`
- [ ] `iirfilter`
- [ ] `kaiser_atten`
- [ ] `kaiser_beta`
- [ ] `kaiserord`
- [ ] `minimum_phase`
- [ ] `savgol_coeffs`
- [ ] `remez`
- [ ] `unique_roots`
- [ ] `residue`
- [ ] `residuez`
- [ ] `invres`
- [ ] `invresz`
- [ ] `BadCoefficients`

**Lower-level filter design functions**
- [ ] `abcd_normalize`
- [ ] `band_stop_obj`
- [ ] `besselap`
- [ ] `buttap`
- [ ] `cheb1ap`
- [ ] `cheb2ap`
- [ ] `cmplx_sort`
- [ ] `ellipap`
- [ ] `lp2bp`
- [ ] `lp2bp_zpk`
- [ ] `lp2bs`
- [ ] `lp2bs_zpk`
- [ ] `lp2hp`
- [ ] `lp2hp_zpk`
- [ ] `lp2lp`
- [ ] `lp2lp_zpk`
- [ ] `normalize`

---

### MATLAB-style IIR filter design
- [ ] `butter`
- [ ] `buttord`
- [ ] `cheby1`
- [ ] `cheb1ord`
- [ ] `cheby2`
- [ ] `cheb2ord`
- [ ] `ellip`
- [ ] `ellipord`
- [ ] `bessel`
- [ ] `iirnotch`
- [ ] `iirpeak`
- [ ] `iircomb`

---

### Continuous-time linear systems
- [ ] `lti`
- [ ] `StateSpace`
- [ ] `TransferFunction`
- [ ] `ZerosPolesGain`
- [ ] `lsim`
- [ ] `lsim2`
- [ ] `impulse`
- [ ] `impulse2`
- [ ] `step`
- [ ] `step2`
- [ ] `freqresp`
- [ ] `bode`

---

### Discrete-time linear systems
- [ ] `dlti`
- [ ] `StateSpace`
- [ ] `TransferFunction`
- [ ] `ZerosPolesGain`
- [ ] `dlsim`
- [ ] `dimpulse`
- [ ] `dstep`
- [ ] `dfreqresp`
- [ ] `dbode`

---

### LTI representations
- [ ] `tf2zpk`
- [ ] `tf2sos`
- [ ] `tf2ss`
- [ ] `zpk2tf`
- [ ] `zpk2sos`
- [ ] `zpk2ss`
- [ ] `ss2tf`
- [ ] `ss2zpk`
- [ ] `sos2zpk`
- [ ] `sos2tf`
- [ ] `cont2discrete`
- [ ] `place_poles`

---

### Waveforms
- [ ] `chirp`
- [ ] `gausspulse`
- [ ] `max_len_seq`
- [ ] `sawtooth`
- [ ] `square`
- [ ] `sweep_poly`
- [ ] `unit_impulse`

---

### Window functions
See 

---

### Wavelets
- [ ] `cascade`
- [ ] `daub`
- [ ] `morlet`
- [ ] `qmf`
- [ ] `ricker`
- [ ] `morlet2`
- [ ] `cwt`

---

### Peak finding
- [ ] `argrelmin`
- [ ] `argrelmax`
- [ ] `argrelextrema`
- [ ] `find_peaks`
- [ ] `find_peaks_cwt`
- [ ] `peak_prominences`
- [ ] `peak_widths`

---

### Spectral analysis
- [ ] `periodogram`
- [ ] `welch`
- [ ] `csd`
- [ ] `coherence`
- [ ] `spectrogram`
- [ ] `lombscargle`
- [ ] `vectorstrength`
- [ ] `stft`
- [ ] `istft`
- [ ] `check_COLA`
- [ ] `check_NOLA`

---

### Chirp Z-transform and Zoom FFT
- [ ] `czt`
- [ ] `zoom_fft`
- [ ] `CZT`
- [ ] `ZoomFFT`
- [ ] `czt_points`

**Comment by leofang** [comment]
cuSignal (https://github.com/rapidsai/cusignal) already implemented quite a lot of these APIs, I'd suggest some coordination (Thoughts, @awthomp?)

**Comment by awthomp** [comment]
My admittedly biased opinion is that we work together to make cusignal better as cusignal already heavily leverages cupy to implement the vast majority of these functions. In many cases, we've optimized performance with elementwise and raw cupy CUDA kernels. Is there a way for `cupy.scipyx.signal` aliases to `cusignal`?

**Comment by ev-br** [comment]
This is exciting! Is there an API reference documentation for cusignal? In addition to the changelog and poking around on GH and in ipython, of course.

**Comment by awthomp** [comment]
Hey @ev-br! You bet! Our documentation for cusignal is here: https://docs.rapids.ai/api/cusignal/stable/

I also tried to show ipynb examples for each API function and compare CPU vs GPU performance. You can see a lot of that work here: https://github.com/rapidsai/cusignal/tree/branch-23.04/notebooks/api_guide

Additonally, we've done some work with SDR integration and E2E deep learning (showing cusignal -> pytorch and cusignal -> nemo for speech to text inferencing).

https://github.com/rapidsai/cusignal/tree/branch-23.04/notebooks/sdr
https://github.com/rapidsai/cusignal/blob/branch-23.04/notebooks/E2E_Example.ipynb

**Comment by ev-br** [comment]
Ah, thank you @awthomp! I did not realize at first that the ipynb examples are meant to be exhaustive or nearly so. Great examples, too!

I agree it would be great to coordinate. To this end, do you have a roadmap for cusignal somewhere, or if not, would you mind sharing the plans for {short, medium, long}-term future of cusignal?

**Comment by takagi** [comment]
@ev-br To make things clearer, I think of labeling these signal functions by the two axes:
- Importance
- Already implemented in cuSignal or not

As to the first point (importance), if possible, do you know a scipy.signal maintainer who could give us some advice about the rating? I mean which functions are more important or widely-used than others generally in a research community.

**Comment by andfoy** [comment]
Just for reference, I've just updated this issue and https://github.com/cupy/cupy/issues/7404 to reflect which APIs are already available in CuSignal. Most of CuSignal available APIs are related to windows, spectral analysis and waveforms/wavelets

**Comment by ev-br** [comment]
As a partial answer for the moment, I think wavelets is scipy.signal are lowest prio: they are effectively superseded by PyWavelets (and cuSignal on GPU, as Edgar points out).

**Comment by awthomp** [comment]
> Just for reference, I've just updated this issue and #7404 to reflect which APIs are already available in CuSignal. Most of CuSignal available APIs are related to windows, spectral analysis and waveforms/wavelets

One reason why we focused cuSignal towards spectral analysis and FIR filtering is that, from my experience, IIR filters are typically lower order and their implementation doesn't really map super well to GPU. We tried to tackle `lfilter` a handful of times, but the DF-II implementation has a data dependency that makes it really tough to parallelize and shine on GPU.

**Comment by takagi** [comment]
@awthomp Did your team make a priority table for these scipy.signal APIs?

Although I’m not sure how we should concretely go (aliasing to cuSignal or having independent codes), we could develop GPU algorithms for the rest of these APIs and share them with cuSignal team.

Here why I said “algorithms” and what prevents us from just aliasing to cuSignal is that we have CuPy specific requirements that we want to also support AMD platform (please don’t feel bad!) and provide conda-free package management.

**Comment by awthomp** [comment]
> @awthomp Did your team make a priority table for these scipy.signal APIs?
> 
> Although I’m not sure how we should concretely go (aliasing to cuSignal or having independent codes), we could develop GPU algorithms for the rest of these APIs and share them with cuSignal team.
> 
> Here why I said “algorithms” and what prevents us from just aliasing to cuSignal is that we have CuPy specific requirements that we want to also support AMD platform (please don’t feel bad!) and provide conda-free package management.

Hey @takagi -- We are discussing this topic internally, this week, and will update the thread in a few days. Thanks!

**Comment by takagi** [comment]
Thanks! I'll wait for the meeting arranged.

**Comment by ev-br** [comment]
NB https://github.com/scipy/scipy/pull/15929 lsim2 is deprecated in scipy, can be ignored here

**Comment by ev-br** [comment]
Just a note that bessel filters, currently blocked on special functions from AMOS, are likely to be unblocked relatively soon by https://github.com/scipy/scipy/pull/19601 and https://github.com/scipy/scipy/pull/19587

**Comment by jakirkham** [comment]
It looks like a few above are merged. Just need there checkbox ticked 😉

Am I reading correctly that there are only 2 left?

**Comment by takagi** [comment]
Yes!
